File: Data Structures & Algorithms/add-two-numbers/submission-2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode ans = new ListNode();
        ListNode cur1= l1;
        ListNode cur2= l2;
        ListNode cur = ans;
        int carry =0;
        while(cur1!=null && cur2!=null)
        {
            int s= cur1.val+cur2.val+carry;
            carry = s/10;
            s= s%10;
            ListNode n  = new ListNode(s);
            if(cur==null)
            {
                cur=n;
            }
            else
            {
                cur.next = n;
                cur =cur.next;
            }
            cur1=cur1.next;
            cur2=cur2.next;
        }
        if(cur1!=null)
        {
            if(carry ==0)
            {
                cur.next = cur1;
            }
            else
            {
                while(carry!=0 && cur1!=null)
                {
                    int s = carry+ cur1.val;
                    carry = s/10;
                    s= s%10;
                    ListNode n  = new ListNode(s);
                    if(cur==null)
                    {
                        cur=n;
                    }
                    else
                    {
                        cur.next = n;
                        cur =cur.next;
                    }
                    cur1=cur1.next;

                }
                if(carry==1)
                {
                    ListNode n  = new ListNode(1);
                    cur.next = n;
                }
                if(cur1!=null)
                {
                    cur.next = cur1;
                }
            }
        }
        if(cur2!=null)
        {
            if(carry ==0)
            {
                cur.next = cur2;
            }
            else
            {
                while(carry!=0 && cur2!=null)
                {
                    int s = carry+ cur2.val;
                    carry = s/10;
                    s= s%10;
                    ListNode n  = new ListNode(s);
                    if(cur==null)
                    {
                        cur=n;
                    }
                    else
                    {
                        cur.next = n;
                        cur =cur.next;
                    }
                    cur2=cur2.next;
                }
                if(carry==1)
                {
                    ListNode n  = new ListNode(1);
                    cur.next = n;
                }
                if(cur2!=null)
                {
                    cur.next = cur2;
                }
            }
        }
        if(carry==1)
        {
            ListNode n  = new ListNode(1);
            cur.next = n;
        }
        return ans.next;
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:

    def findMid(self,head):
        slow = head
        fast = head
        
        while fast and fast.next:
            
            fast = fast.next.next
            slow = slow.next

        return slow

    def reverse(self,head):
        prev = None
        curr = head

        while curr:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp
        
        return prev

    def reorder(self,head,head1):
        cur = head
        cur2 = head1
                
        while cur and cur2:
            temp = cur.next
            temp2 = cur2.next
            cur.next = cur2
            cur2.next = temp

            cur = temp
            cur2 = temp2

        

        return head

    def reorderList(self, head: Optional[ListNode]) -> None:
        if not head or not head.next:
            return
        mid = self.findMid(head)

        #print(mid.val)
        m1 = mid.next

        mid.next = None
        m2 = self.reverse(m1)

        self.reorder(head,m2)